### ISSUE TITLE
Verify everything works correctly with scrollTopLeftInterop mode enabled

### ISSUE THREAD

**Issue by RByers** [gblazex/smoothscroll#111]
We're finally [fixing the scrollTop bug in Chrome](https://dev.opera.com/articles/fixing-the-scrolltop-bug/).  This code appears broken:

``` javascript
function getScrollRoot() {
    return document.body; // scrolling root in WebKit
}
```

But when I tested the extension appeared to work fine still due to this:

``` javascript
var scrollWindow = (elem === document.body);
...
if (scrollWindow) {
   window.scrollBy(scrollX, scrollY);
} 
```

So I don't have any particular reason to believe SmoothScroll will break when we make `scrollTopLeftInterop` mode [the default](http://crbug.com/157855) (hopefully in Chrome 47).  Still, since the code is suspect you might want to test to be sure.

I came across this when investigating [a bug in a website](https://code.google.com/p/chromium/issues/detail?id=501568) which appears to have copied an older version of your code (which was in fact broken).

**Comment by galambalazs** [comment]
Thanks a lot for the heads up Rick! 

It's good to see someone from the Chrome team caring about smooth scrolling. :-) 
(even if just for compatibility)
As you mentioned in Issue 501568, you guys did leave SS unfixed for tooo long.
I don't like the current implementation either. There's no acceleration/momentum.

Can you ping someone at the Chrome team so we could cook something together?
I'd love to see proper smooth scrolling in Chrome.


I don't worry about my extension's compatibility because that can be auto-updated in a heartbeat. The bigger issue is the sites that use the embedded version as you mentioned.

We can't get them to update easily (even if a WP theme gets updated, the individual websites using it may not).

I made the mistake of publishing the Website version of SS as a gist. Then it got really popular: https://gist.github.com/galambalazs/6477177

Lots of sites unfortunately copied the code and never updated it. That's why I recently created a separate repo for that branch with bower support: https://github.com/galambalazs/smoothscroll-for-websites



You are correct that `document.body` is only used to "label" when the root is scrolled, but the actual scrolling is done by `window.scrollBy`

The up-to-date Website version even has this check just to make sure it works everywhere  (although this might be an overkill to send over the wire, if `window.scrollBy` sidesteps the problem anyway).

```
var getScrollRoot = (function() {
  var SCROLL_ROOT;
  return function() {
    if (!SCROLL_ROOT) {
      var dummy = document.createElement('div');
      dummy.style.cssText = 'height:10000px;width:1px;';
      document.body.appendChild(dummy);
      var bodyScrollTop  = document.body.scrollTop;
      var docElScrollTop = document.documentElement.scrollTop;
      window.scrollBy(0, 1);
      if (document.body.scrollTop != bodyScrollTop)
        (SCROLL_ROOT = document.body);
      else 
        (SCROLL_ROOT = document.documentElement);
      window.scrollBy(0, -1);
      document.body.removeChild(dummy);
    }
    return SCROLL_ROOT;
  };
})();
```

**Comment by RByers** [comment]
Great, thanks for the detailed response Balázs!  

> As you mentioned in Issue 501568, you guys did leave SS unfixed for tooo long.
> I don't like the current implementation either. There's no acceleration/momentum.
> Can you ping someone at the Chrome team so we could cook something together?
> I'd love to see proper smooth scrolling in Chrome

We have been working on smooth scrolling recently, but it's been relatively low priority.  I am going to fix that, though possibly not until Q4.  We'd love to get your feedback.  Perhaps the best way to discuss general feedback is for you to start an e-mail thread on input-dev@chromium.org, and I'll make sure the right engineers are looped in.  For concrete specific issues, feel free to [file bugs](http://crbug.com) and I can make sure they're routed to the right people.

That said, I suspect the first version we ship will still be relatively simple and there will probably always be a demand for more powerful / customizable extensions like yours.  We're also working on [making scrolling more customizable](https://docs.google.com/document/d/1VnvAqeWFG9JFZfgG5evBqrLGDZYRE5w6G5jEDORekPY) which should help make extentions/libraries like yours easier to build reliably and performantly.  One big drawback today is that `wheel` handlers necessarily disable our threaded scrolling optimization, which can (depend on the page and device) have a huge negative impact on scroll performance.

One possible path forward for us with `scrollTopLeftInterop` is to wait until we've shipped a decent smooth scrolling implementation in Chrome and then encourage websites to fix the issue with scrolling being broken by removing their usage of `smoothscroll.js` entirely.  Sound OK to you?

> The up-to-date Website version even has this check just to make sure it works everywhere (although this might be an overkill to send over the wire, if window.scrollBy sidesteps the problem anyway).

Yeah if you're using `window.scrollBy` then perhaps you don't really care which element represents the document scroller.  The main disadvantage to the test you're doing is that it may cause a visual jump and unnecessary `scroll` events.  You can fix that by using the new `[Document.scrollingElement](https://drafts.csswg.org/cssom-view/#dom-document-scrollingelement)` API when it's available.

I notice the new website version supports all browsers (where in practice the sites I've found install it only for Chrome).  So even if the browser supports smooth scroll itself, you override that with your own, right?

**Comment by galambalazs** [comment]
> That said, I suspect the first version we ship will still be relatively simple and there will probably always be a demand for more powerful / customizable extensions like yours. We're also working on making scrolling more customizable which should help make extentions/libraries like yours easier to build reliably and performantly. One big drawback today is that wheel handlers necessarily disable our threaded scrolling optimization, which can (depend on the page and device) have a huge negative impact on scroll performance.

I applaud your efforts but I don't think it solves the problem.
Acceleration *is* simple, you just take the input frequency and scale up the scrolling deltas.
Customization is a nice thing but I'd prefer if "it just worked". I mean how many people do you think have the urge to customize how their Mac touchpad scrolls? Less than 1% I'm sure. Because it feels right out of the box. I'd like if Chrome wheel scrolling would hit that 99% utility/UX curve. 

> One possible path forward for us with scrollTopLeftInterop is to wait until we've shipped a decent smooth scrolling implementation in Chrome and then encourage websites to fix the issue with scrolling being broken by removing their usage of smoothscroll.js entirely. Sound OK to you?

Sure, I'm happy to deprecate that version once Chrome ships with smooth scrolling.

**Comment by galambalazs** [comment]
As for the User Agent check, thanks for reminding me. I needed to add that back for Chrome & Safari.

https://github.com/galambalazs/smoothscroll-for-websites/commit/9afaf2276600d2f09e626a0bdbe9c5db7ccefacf

**Comment by RByers** [comment]
> I applaud your efforts but I don't think it solves the problem.
Acceleration is simple, you just take the input frequency and scale up the scrolling deltas.
Customization is a nice thing but I'd prefer if "it just worked". I mean how many people do you think have the urge to customize how their Mac touchpad scrolls? Less than 1% I'm sure. Because it feels right out of the box. I'd like if Chrome wheel scrolling would hit that 99% utility/UX curve.

Yes, the purpose of scroll customization is not for users to enable scenarios like this.  It's for developers to be able to build richer scroll-coupled effects that perform well (I was't suggesting it is designed to help enable smooth scrolling, just that it might be handy for your extension if you continue to feel the need for it).

But yes, I absolutely agree that the goal should be to make Chrome behave great by default so there is no demand for extensions like yours :-).
